policies/testing.md

```diff
@@ -0,0 +1,71 @@
+Testing Policy
+==============
+
+This applies to all stable and development branches of the main code repository.
+
+Within this policy "functional behaviour" means what the system does, i.e.
+given a set of inputs it will produce a set of outputs. This does not include
+how the system does it. For example refactoring or performance improvements do
+not affect functional behaviour.
+
+Except where noted below:
+
+- All Pull Requests adding new functionality to the applications, libraries,
+providers or engines must include suitable tests.
+- All Pull Requests fixing a functional behavior defect in the applications,
```

Comment:
I think it is overly strict to require a test for every small bug fix. Especially for the cases where writing the test would require adding significant infrastructure (tests depending on triggering a malloc failure, tests having to add a new provider implementation or other cases). IMO the policy should be at least somewhat more lenient such as allowing the reviewer(s) to give an exception for this rule.

Reply:
Ah, I see you've added this already below. OK then.

Reply:
I've only added an exception where the reviewers can agree for "difficult to reproduce error conditions". Should there be a broader exception?

Reply:
Maybe add something like that reviewers can give an exception where writing the test would require significantly higher effort than writing the bug fix.

Reply:
I'm not sure "signficiantly higher" is quite the right bar. Its not that uncommon that writing the test is significantly more work than fixing the bug. I don't think that should necessarily be an excuse to not write a test. However, I wouldn't expect someone to have to go and write a whole new testing framework to test specific classes of things. I've made an update that says "disproportionately more effort".